### PR TITLE
fix: if file is new then unstage it before discarding, use clean command instead of fs for removing untracked file

### PR DESCRIPTION
**Pull request by golden-expiriensu**
Fixes: https://github.com/altsem/gitu/issues/114

### CODE REVIEW

**Comment by altsem** [comment]
Nice! I'll have a look :)

**Review by altsem — COMMENTED** [review]
Good to see some love for these. I did some testing and seems to work!
Is `git rm --force *file*` a shorter way to achieve the same thing?

File: src/ops/discard.rs [hunk 1]
```diff
@@ -10,20 +10,22 @@ impl OpTrait for Discard {
...
+                    git2::Delta::Added => Rc::new(move |state: &mut State, term: &mut Term| {
+                        let file = d.new_file.as_os_str();
+                        state.run_cmd(term, &[], unstage_file_cmd(file))?;
+                        state.run_cmd(term, &[], clean_file_cmd(file))
+                    }),
+                    _ => {
+                        if d.old_file == d.new_file {
+                            cmd_arg(checkout_file_cmd, d.old_file.into())
+                        } else {
+                            // TODO Discard file move
```
Comment (altsem) [inline src/ops/discard.rs:25]: Optional: I think this old_file == new_file check and TODO can just be removed. I can't recall it ever happening, and should now be covered by your match on line 15 anyway!

File: src/ops/discard.rs [hunk 2]
```diff
@@ -10,20 +10,22 @@ impl OpTrait for Discard {
...
             Some(TargetData::Branch(r)) => cmd_arg(discard_branch, r.into()),
-            Some(TargetData::File(f)) => Rc::new(move |state: &mut State, _term: &mut Term| {
-                let path = PathBuf::from_iter([
-                    state.repo.workdir().expect("No workdir").to_path_buf(),
-                    f.clone(),
-                ]);
-                std::fs::remove_file(path)?;
-                state.screen_mut().update()
-            }),
+            Some(TargetData::File(f)) => cmd_arg(clean_file_cmd, f.into()),
```
Comment (altsem) [inline src/ops/discard.rs:13]: Optional: I had a look at why there's TargetData::File and TargetData::Delta, and found some refactoring opportunity. It should be possible to remove TargetData::File completely removed from the code-base!

Comment (golden-expiriensu) [reply src/ops/discard.rs:13]: There is one place where TargetData::File is created and it is in items_list. I tried to replace it with delta but got confused pretty quick. I think I need to look into that one more time later...

**Comment by golden-expiriensu** [comment]
> Good to see some love for these. I did some testing and seems to work! Is `git rm --force *file*` a shorter way to achieve the same thing? i.e.:
> 
> * added untracked (git clean --force _file_)
> * added unstaged (git rm --force _file_)
> * added staged (git rm --force _file_)

This is a beautiful solution, thank you!

File: src/ops/discard.rs (hunk 1 above)
Comment (golden-expiriensu) [reply src/ops/discard.rs]: Done! Is it intentional that it shows moved file as a pair of added-deleted?

Comment (altsem) [reply src/ops/discard.rs]: I think there must be some configuration to do move detection with libgit2.
It doesn't seem to happen out of default. Just a matter of implementing/fixing

**Review by altsem — APPROVED** [review]
I'll squash these with a more user-friendly message as the "fix: " commit is the one that'll appear in the changelog.
:rocket:

**Pull request merged**